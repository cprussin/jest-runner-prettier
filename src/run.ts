import { TestResult } from "@jest/test-result";
import { emphasize } from "emphasize";
import { pass, fail } from "create-lite-jest-runner";
import * as fs from "node:fs/promises";
import { diff } from "jest-diff";
import prettier from "prettier";

interface Parameters {
  testPath: string;
  ignorePath?: string;
}

export default async ({ testPath, ignorePath }: Parameters): Promise<TestResult> => {
  const start = Date.now();
  const fileInfo = await prettier.getFileInfo(testPath, { ignorePath });

  if (fileInfo.ignored) {
    const end = Date.now();
    return {
      leaks: false,
      numFailingTests: 0,
      numPassingTests: 0,
      numPendingTests: 1,
      numTodoTests: 0,
      openHandles: [],
      perfStats: {
        start,
        end,
        runtime: end - start,
        slow: false,
      },
      skipped: true,
      snapshot: {
        added: 0,
        fileDeleted: false,
        matched: 0,
        unchecked: 0,
        uncheckedKeys: [],
        unmatched: 0,
        updated: 0,
      },
      testFilePath: testPath,
      testResults: [],
    };
  }

  const contents = await fs.readFile(testPath, "utf8");
  const config = await prettier.resolveConfig(testPath);

  const prettierConfig = {
    ...config,
    filepath: testPath,
  };

  const isPretty = prettier.check(contents, prettierConfig);
  if (isPretty) {
    return pass({
      start,
      end: Date.now(),
      test: { path: testPath },
    });
  }

  const formatted = prettier.format(contents, prettierConfig);

  return fail({
    start,
    end: Date.now(),
    test: {
      path: testPath,
      errorMessage: diff(
        emphasize.highlightAuto(formatted).value,
        emphasize.highlightAuto(contents).value,
        {
          expand: false,
        }
      ) as string | undefined,
    },
  });
};
