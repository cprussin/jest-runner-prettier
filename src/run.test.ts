import path from "node:path";
import run from "./run";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Remove undeterministic data from test reports
expect.addSnapshotSerializer({
  print: (value: any, serialize) => {
    delete value.perfStats;
    delete value.testFilePath;
    for (const result of value.testResults) {
      delete result.duration;
    }
    return serialize(value);
  },
  test: (value) =>
    value && value.perfStats && value.testFilePath && value.testResults,
});

describe("jest-runner-prettier", () => {
  describe("JSON", () => {
    describe("good fixture", () => {
      it("matches snapshot", () => {
        return run({
          testPath: path.join(__dirname, "__fixtures__", `good.json`),
        }).then((result) => expect(result).toMatchSnapshot());
      });
    });

    describe("bad fixture", () => {
      it("matches snapshot", () => {
        return run({
          testPath: path.join(__dirname, "__fixtures__", `bad.json`),
        }).then((result) => expect(result).toMatchSnapshot());
      });
    });
  });

  describe("JSX", () => {
    describe("good fixture", () => {
      it("matches snapshot", () => {
        return run({
          testPath: path.join(__dirname, "__fixtures__", `good.jsx`),
        }).then((result) => expect(result).toMatchSnapshot());
      });
    });

    describe("bad fixture", () => {
      it("matches snapshot", () => {
        return run({
          testPath: path.join(__dirname, "__fixtures__", `bad.jsx`),
        }).then((result) => expect(result).toMatchSnapshot());
      });
    });
  });
});
