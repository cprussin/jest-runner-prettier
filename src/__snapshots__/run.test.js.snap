// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest-runner-prettier JSON bad fixture matches snapshot 1`] = `
Object {
  "console": null,
  "failureMessage": "[32m- Expected[39m
[31m+ Received[39m

[32m- {[39m
[32m-   [36m\\"ohMyGodLongArray\\"[32m: [[39m
[32m-     [31m\\"prettyLongValue\\"[32m,[39m
[32m-     [31m\\"decentlyLongValue\\"[32m,[39m
[32m-     [31m\\"prettyDecentlyLongValue\\"[32m,[39m
[32m-     [31m\\"wowWowWowWhatALongValue\\"[32m[39m
[32m-   ],[39m
[32m-   [36m\\"woahThisObjectIsDisapperingOffTheScreen\\"[32m: {[39m
[32m-     [36m\\"foo\\"[32m: [31m\\"bar\\"[32m,[39m
[32m-     [36m\\"bar\\"[32m: [31m\\"foo\\"[32m,[39m
[32m-     [36m\\"faz\\"[32m: [34mfalse[32m[39m
[32m-   },[39m
[32m-   [36m\\"truthy\\"[32m: [34mtrue[32m,[39m
[32m-   [36m\\"aNumber\\"[32m: [32m123[32m,[39m
[32m-   [36m\\"aString\\"[32m: [31m\\"hi\\"[32m[39m
[32m- }[39m
[32m- [39m
[31m+ {[36m\\"ohMyGodLongArray\\"[31m: [[31m\\"prettyLongValue\\"[31m, [31m\\"decentlyLongValue\\"[31m,[39m
[31m+   [31m\\"prettyDecentlyLongValue\\"[31m, [31m\\"wowWowWowWhatALongValue\\"[31m], [36m\\"woahThisObjectIsDisapperingOffTheScreen\\"[31m: {[36m\\"foo\\"[31m: [31m\\"bar\\"[31m,[39m
[31m+     [36m\\"bar\\"[31m: [31m\\"foo\\"[31m, [36m\\"faz\\"[31m: [34mfalse[31m}, [36m\\"truthy\\"[31m: [34mtrue[31m, [36m\\"aNumber\\"[31m: [32m123[31m, [36m\\"aString\\"[31m: [31m\\"hi\\"[31m[39m
[31m+   }[39m",
  "numFailingTests": 1,
  "numPassingTests": 0,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "skipped": undefined,
  "snapshot": Object {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "unmatched": 0,
    "updated": 0,
  },
  "sourceMaps": Object {},
  "testExecError": null,
  "testResults": Array [
    Object {
      "ancestorTitles": Array [],
      "failureMessages": Array [
        "[32m- Expected[39m
[31m+ Received[39m

[32m- {[39m
[32m-   [36m\\"ohMyGodLongArray\\"[32m: [[39m
[32m-     [31m\\"prettyLongValue\\"[32m,[39m
[32m-     [31m\\"decentlyLongValue\\"[32m,[39m
[32m-     [31m\\"prettyDecentlyLongValue\\"[32m,[39m
[32m-     [31m\\"wowWowWowWhatALongValue\\"[32m[39m
[32m-   ],[39m
[32m-   [36m\\"woahThisObjectIsDisapperingOffTheScreen\\"[32m: {[39m
[32m-     [36m\\"foo\\"[32m: [31m\\"bar\\"[32m,[39m
[32m-     [36m\\"bar\\"[32m: [31m\\"foo\\"[32m,[39m
[32m-     [36m\\"faz\\"[32m: [34mfalse[32m[39m
[32m-   },[39m
[32m-   [36m\\"truthy\\"[32m: [34mtrue[32m,[39m
[32m-   [36m\\"aNumber\\"[32m: [32m123[32m,[39m
[32m-   [36m\\"aString\\"[32m: [31m\\"hi\\"[32m[39m
[32m- }[39m
[32m- [39m
[31m+ {[36m\\"ohMyGodLongArray\\"[31m: [[31m\\"prettyLongValue\\"[31m, [31m\\"decentlyLongValue\\"[31m,[39m
[31m+   [31m\\"prettyDecentlyLongValue\\"[31m, [31m\\"wowWowWowWhatALongValue\\"[31m], [36m\\"woahThisObjectIsDisapperingOffTheScreen\\"[31m: {[36m\\"foo\\"[31m: [31m\\"bar\\"[31m,[39m
[31m+     [36m\\"bar\\"[31m: [31m\\"foo\\"[31m, [36m\\"faz\\"[31m: [34mfalse[31m}, [36m\\"truthy\\"[31m: [34mtrue[31m, [36m\\"aNumber\\"[31m: [32m123[31m, [36m\\"aString\\"[31m: [31m\\"hi\\"[31m[39m
[31m+   }[39m",
      ],
      "fullName": undefined,
      "numPassingAsserts": 1,
      "status": "failed",
      "title": "",
    },
  ],
}
`;

exports[`jest-runner-prettier JSON good fixture matches snapshot 1`] = `
Object {
  "console": null,
  "failureMessage": undefined,
  "numFailingTests": 0,
  "numPassingTests": 1,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "skipped": undefined,
  "snapshot": Object {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "unmatched": 0,
    "updated": 0,
  },
  "sourceMaps": Object {},
  "testExecError": null,
  "testResults": Array [
    Object {
      "ancestorTitles": Array [],
      "failureMessages": Array [
        undefined,
      ],
      "fullName": undefined,
      "numPassingAsserts": 0,
      "status": "passed",
      "title": "",
    },
  ],
}
`;

exports[`jest-runner-prettier JSX bad fixture matches snapshot 1`] = `
Object {
  "console": null,
  "failureMessage": "[32m- Expected[39m
[31m+ Received[39m

[32m- [33m[34mfunction[33m HelloWorld({[32m[39m
[32m- [33m  greeting = [31m\\"hello\\"[33m,[32m[39m
[32m- [33m  greeted = [31m'\\"World\\"'[33m,[32m[39m
[32m- [33m  silent = false,[32m[39m
[32m- [33m  onMouseOver[32m[39m
[32m- [33m}) [32m{[39m
[32m-   [34mif[32m (!greeting) {[39m
[32m-     [34mreturn[32m [34mnull[32m;[39m
[32m-   }[39m
[31m+ [33m[34mfunction[33m HelloWorld({greeting = [31m\\"hello\\"[33m, greeted = [31m'\\"World\\"'[33m, silent = false, onMouseOver,}) [31m{[39m
[31m+ [39m
[31m+   [34mif[31m(!greeting){[34mreturn[31m [34mnull[31m};[39m
[31m+ [39m
[31m+      [32m// [32mTODO:[32m Don't use random in render[31m[39m
[31m+   [34mlet[31m num = [36mMath[31m.floor ([36mMath[31m.random() * [32m1E+7[31m).toString().replace([31m/\\\\.\\\\d+/ig[31m, [31m\\"\\"[31m)[39m
[31m+ [39m
[31m+   [34mreturn[31m [90m<[34mdiv[90m [36mclassName[90m=[31m'HelloWorld'[90m [36mtitle[90m=[31m{[90m\`[36mYou[90m [36mare[90m [36mvisitor[90m [36mnumber[90m \${ [36mnum[90m }\`} [36monMouseOver[90m=[31m{onMouseOver}[90m>[31m[39m
[31m+ [39m
[31m+     [90m<[34mstrong[90m>[31m{ greeting.slice( 0, 1 ).toUpperCase() + greeting.slice(1).toLowerCase() }[90m</[34mstrong[90m>[31m[39m
[31m+     {greeting.endsWith(\\",\\") ? \\" \\" : [90m<[34mspan[90m [36mstyle[90m=[31m{{color:[90m '\\\\[36mgrey[90m'}}>[31m\\", \\"[90m</[34mspan[90m>[31m }[39m
[31m+     [90m<[34mem[90m>[31m[39m
[31m+ 	{ greeted }[39m
[31m+ 	[90m</[34mem[90m>[31m[39m
[31m+     { (silent)[39m
[31m+       ? \\".\\"[39m
[31m+       : \\"!\\"}[39m
[2m  [22m
[32m-   [32m// [32mTODO:[32m Don't use random in render[32m[39m
[32m-   [34mlet[32m num = [36mMath[32m.floor([36mMath[32m.random() * [32m1e7[32m)[39m
[32m-     .toString()[39m
[32m-     .replace([31m/\\\\.\\\\d+/gi[32m, [31m\\"\\"[32m);[39m
[31m+     [90m</[34mdiv[90m>[31m;[39m
[2m  [22m
[32m-   [34mreturn[32m ([39m
[32m-     [90m<[34mdiv[90m[32m[39m
[32m- [90m      [36mclassName[90m=[31m\\"HelloWorld\\"[90m[32m[39m
[32m- [90m      [36mtitle[90m=[31m{[90m\`[36mYou[90m [36mare[90m [36mvisitor[90m [36mnumber[90m \${[36mnum[90m}\`}[32m[39m
[32m- [90m      [36monMouseOver[90m=[31m{onMouseOver}[90m[32m[39m
[32m- [90m    >[32m[39m
[32m-       [90m<[34mstrong[90m>[32m[39m
[32m-         {greeting.slice(0, 1).toUpperCase() + greeting.slice(1).toLowerCase()}[39m
[32m-       [90m</[34mstrong[90m>[32m[39m
[32m-       {greeting.endsWith(\\",\\") ? ([39m
[32m-         \\" \\"[39m
[32m-       ) : ([39m
[32m-         [90m<[34mspan[90m [36mstyle[90m=[31m{{[90m [36mcolor:[90m \\"[36mgrey[90m\\" }}>[32m\\", \\"[90m</[34mspan[90m>[32m[39m
[32m-       )}[39m
[32m-       [90m<[34mem[90m>[32m{greeted}[90m</[34mem[90m>[32m[39m
[32m-       {silent ? \\".\\" : \\"!\\"}[39m
[32m-     [90m</[34mdiv[90m>[32m[39m
[32m-   );[39m
[2m  }[22m
[32m- [39m",
  "numFailingTests": 1,
  "numPassingTests": 0,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "skipped": undefined,
  "snapshot": Object {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "unmatched": 0,
    "updated": 0,
  },
  "sourceMaps": Object {},
  "testExecError": null,
  "testResults": Array [
    Object {
      "ancestorTitles": Array [],
      "failureMessages": Array [
        "[32m- Expected[39m
[31m+ Received[39m

[32m- [33m[34mfunction[33m HelloWorld({[32m[39m
[32m- [33m  greeting = [31m\\"hello\\"[33m,[32m[39m
[32m- [33m  greeted = [31m'\\"World\\"'[33m,[32m[39m
[32m- [33m  silent = false,[32m[39m
[32m- [33m  onMouseOver[32m[39m
[32m- [33m}) [32m{[39m
[32m-   [34mif[32m (!greeting) {[39m
[32m-     [34mreturn[32m [34mnull[32m;[39m
[32m-   }[39m
[31m+ [33m[34mfunction[33m HelloWorld({greeting = [31m\\"hello\\"[33m, greeted = [31m'\\"World\\"'[33m, silent = false, onMouseOver,}) [31m{[39m
[31m+ [39m
[31m+   [34mif[31m(!greeting){[34mreturn[31m [34mnull[31m};[39m
[31m+ [39m
[31m+      [32m// [32mTODO:[32m Don't use random in render[31m[39m
[31m+   [34mlet[31m num = [36mMath[31m.floor ([36mMath[31m.random() * [32m1E+7[31m).toString().replace([31m/\\\\.\\\\d+/ig[31m, [31m\\"\\"[31m)[39m
[31m+ [39m
[31m+   [34mreturn[31m [90m<[34mdiv[90m [36mclassName[90m=[31m'HelloWorld'[90m [36mtitle[90m=[31m{[90m\`[36mYou[90m [36mare[90m [36mvisitor[90m [36mnumber[90m \${ [36mnum[90m }\`} [36monMouseOver[90m=[31m{onMouseOver}[90m>[31m[39m
[31m+ [39m
[31m+     [90m<[34mstrong[90m>[31m{ greeting.slice( 0, 1 ).toUpperCase() + greeting.slice(1).toLowerCase() }[90m</[34mstrong[90m>[31m[39m
[31m+     {greeting.endsWith(\\",\\") ? \\" \\" : [90m<[34mspan[90m [36mstyle[90m=[31m{{color:[90m '\\\\[36mgrey[90m'}}>[31m\\", \\"[90m</[34mspan[90m>[31m }[39m
[31m+     [90m<[34mem[90m>[31m[39m
[31m+ 	{ greeted }[39m
[31m+ 	[90m</[34mem[90m>[31m[39m
[31m+     { (silent)[39m
[31m+       ? \\".\\"[39m
[31m+       : \\"!\\"}[39m
[2m  [22m
[32m-   [32m// [32mTODO:[32m Don't use random in render[32m[39m
[32m-   [34mlet[32m num = [36mMath[32m.floor([36mMath[32m.random() * [32m1e7[32m)[39m
[32m-     .toString()[39m
[32m-     .replace([31m/\\\\.\\\\d+/gi[32m, [31m\\"\\"[32m);[39m
[31m+     [90m</[34mdiv[90m>[31m;[39m
[2m  [22m
[32m-   [34mreturn[32m ([39m
[32m-     [90m<[34mdiv[90m[32m[39m
[32m- [90m      [36mclassName[90m=[31m\\"HelloWorld\\"[90m[32m[39m
[32m- [90m      [36mtitle[90m=[31m{[90m\`[36mYou[90m [36mare[90m [36mvisitor[90m [36mnumber[90m \${[36mnum[90m}\`}[32m[39m
[32m- [90m      [36monMouseOver[90m=[31m{onMouseOver}[90m[32m[39m
[32m- [90m    >[32m[39m
[32m-       [90m<[34mstrong[90m>[32m[39m
[32m-         {greeting.slice(0, 1).toUpperCase() + greeting.slice(1).toLowerCase()}[39m
[32m-       [90m</[34mstrong[90m>[32m[39m
[32m-       {greeting.endsWith(\\",\\") ? ([39m
[32m-         \\" \\"[39m
[32m-       ) : ([39m
[32m-         [90m<[34mspan[90m [36mstyle[90m=[31m{{[90m [36mcolor:[90m \\"[36mgrey[90m\\" }}>[32m\\", \\"[90m</[34mspan[90m>[32m[39m
[32m-       )}[39m
[32m-       [90m<[34mem[90m>[32m{greeted}[90m</[34mem[90m>[32m[39m
[32m-       {silent ? \\".\\" : \\"!\\"}[39m
[32m-     [90m</[34mdiv[90m>[32m[39m
[32m-   );[39m
[2m  }[22m
[32m- [39m",
      ],
      "fullName": undefined,
      "numPassingAsserts": 1,
      "status": "failed",
      "title": "",
    },
  ],
}
`;

exports[`jest-runner-prettier JSX good fixture matches snapshot 1`] = `
Object {
  "console": null,
  "failureMessage": undefined,
  "numFailingTests": 0,
  "numPassingTests": 1,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "skipped": undefined,
  "snapshot": Object {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "unmatched": 0,
    "updated": 0,
  },
  "sourceMaps": Object {},
  "testExecError": null,
  "testResults": Array [
    Object {
      "ancestorTitles": Array [],
      "failureMessages": Array [
        undefined,
      ],
      "fullName": undefined,
      "numPassingAsserts": 0,
      "status": "passed",
      "title": "",
    },
  ],
}
`;
